README.md
# PiProxy

*Reverse Proxy in NodeJS for HTTP/HTTPS/HTTP2 with Automatic SSL Management, Compression, Security Enforcement and Rate Limiting*

When you run a server app that runs as web server, you want a way to expose it using user-friendly URL  
And typically this is done using and configuring a commerical grade server such as [nginx](https://nginx.org/en/) that will perform reverse proxy functions from user-friendly URL to your local server  

But what if you want to do it all in `NodeJS`?  
And have automatic handling of dynamic DNS and SSL certificates?  

<br>

## Features

- Native HTTP2 support as front-end
- Can proxy to HTTP2, HTTPS or HTTP destinations
- ACME/LetsEncrypt support for automatic creation and renewal of free, valid and signed SSL certificates
- No-IP support for automatic dynamic IP updates such as ddns.net and others
- Configurable passthrough compression using Brotli algorithm
- Optional rate limiting using sliding window
- TLS version protection (TLS v1.2 and higher are allowed)
- Helmet and CSP protection
- GeoIP reverse lookups on access (large package size is due to GeoIP lite databases included in the package)
- Custom error handling
- Agent analysis on access
- Full logging support

<br>

## Install & Run

### Prerequisites

1. Install [NodeJS](https://nodejs.org/en/)  
2. Set [NodeJS](https://nodejs.org/en/) permissions  
   See section [Permissions](###Permissions) for options  
3. Choose how you want to handle SSL certificates  
   See section [SSL Options](###SSL-Options) for options  
4. Set your **router** to redirect all HTTPS traffic to your proxy server
   This will allow secure access from internet  
   Alternatively, run without external access  
   Note that external access is required to require valid signed SSL certificate  
5. If you want to perform optional location lookups based on IP,  
   download [MaxMind](https://dev.maxmind.com/geoip/geolite2-free-geolocation-data?lang=en) city and ASN databases

### Install & Run

1. Clone latest version of `piproxy`
    > git clone --depth 1 https://github.com/vladmandic/piproxy`  
2. Create configuration  
   Rename `config.json.sample` to `config.json` and modify it as needed  
   See section [Configuration](##Configuration) for details
3. Start `piproxy`:
    > `npm start`  

    or use `piproxy.service` as a template (notes are within service file) to create a Linux `systemd` service  

<br>

## Configuration

Example of minimum configuration from `config.json.sample`:

```json
{
  // main section that controls all reverse proxy functions
  "redirects": [
    // for any request matching `weather.ddns.net` fetch data from `locahost:10000`
    { "url": "weather.ddns.net", "target": "localhost", "port": "10000" },
    { "url": "clock.ddns.net", "target": "localhost", "port": "10010" },
    { "url": "profile.ddns.net", "target": "localhost", "port": "10020" },
    // for any request that does not match any rule fetch data from `locahost:10000`
    { "default": true, "target": "localhost", "port": "10000" }
  ],
  // optional: update dynamic ip for a listed hosts using provided username and password at no-ip service
  "noip": {
    "host": [ "weather.ddns.net", "clock.ddns.net", "profile.ddns.net" ],
    "user": "me@example.com",
    "password": "password"
  },
  // optional: automatically handle ssl certificates for listsed domains using let's encrypt service
  "acme": {
    "domains": [ "weather.ddns.net", "clock.ddns.net", "profile.ddns.net" ],
    "maintainer": "me@example.com",
    "subscriber": "me@example.com"
  },
  // optional: provide predefined http responses on some well known static paths
  "answers": {
    "security.txt": "Contact: mailto:me@example.com\nPreferred-Languages: en\n"
  }
}
```

Default locations (can be changed in advanced configuration):
- Certificates are stored in `/cert`
- Logs are stored in `/logs`
- GeoIP databases are stored in `/geoip`

Full configuration details can be seen at [https://github.com/vladmandic/piproxy/tree/master/src/config.ts]

<br>

## Notes
### SSL Options

- Provide your own private key and server certificate 

- Generate self-signed private key and server certificate

  > openssl req -x509 -newkey rsa:4096 -nodes -days 365 \
    -keyout https.key -out https.crt \
    -subj "/C=US/ST=Florida/L=Miami/O=@vladmandic"

- Allow `piproxy` to automatically get signed certificate from [Let's Encypt](https://letsencrypt.org/)  
  In this case, `piproxy` will automatically handle certificate renewals

### Permissions

To allow the server to listen on privledged ports (TCP ports below 1024), you need to either:

- Configure your system so node process can bind to priviledged ports without running as root:  
  (you only need to do this once)
  > sudo setcap 'cap_net_bind_service=+ep' `which node`

- Run node as root with sudo:
  > sudo node server/proxy.js
  In that case, proxy will automatically try to drop priviledges as soon as server is started

- Run node on a non-priviledged port and  
  forward a desired priviledged port to a non-priviledged one using firewall NAT configuration  
  For example, to route TCP port 443 (https) to port 8000:

  > sudo iptables -t nat -A PREROUTING -i eth0 -p tcp --dport 80 -j REDIRECT --to-port 8000  

  or  

  > sudo ipfw add 100 fwd 127.0.0.1,8000 tcp from any to any 80 in


### Compression

proxy uses Brotli algorithm with a configurable level of compression.
Example of compression efficiency on pure html:

- L0: 2.1 MB (original without compression)
- L1: 692 kB
- L3: 577 kB
- L5: 516 kB
- L9: 494 kB

### GeoIP

GeoIP databases are not included in the package and should be provided by user.
Once databases are available, they can be specified in the configuration:

```js
  // use geoip reverse lookups to locate requests
  // databases can be obtained for free from maxmind at <https://dev.maxmind.com/geoip/geoip2/geolite2/>
  geoIP: {
    city: './geoip/GeoLite2-City.mmdb',
    asn: './geoip/GeoLite2-ASN.mmdb',
  },
```

## Example Log

### Startup

```js
2022-08-14 11:01:13 INFO:  configuration { file: 'config.json' }
2022-08-14 11:01:13 INFO:  { application: '@vladmandic/piproxy', version: '2.0.0' }
2022-08-14 11:01:13 INFO:  { user: 'vlado', platform: 'linux', arch: 'x64', node: 'v18.7.0' }
2022-08-14 11:01:13 STATE: { log: '/home/vlado/dev/piproxy/logs/proxy.log', access: '/home/vlado/dev/piproxy', client: '/home/vlado/dev/piproxy' }
2022-08-14 11:01:13 INFO:  ssl { key: '../cert/private.pem', crt: '../cert/fullchain.pem' }
2022-08-14 11:01:13 INFO:  ssl account { contact: 'mailto:mandic00@live.com', status: 'valid', type: 'EC', crv: 'P-256' }
2022-08-14 11:01:13 INFO:  ssl server { subject: 'pidash.ddns.net', issuer: "C=US, O=Let's Encrypt, CN=R3", algorithm: 'sha256WithRSAEncryption', from: 2022-07-12T12:09:02.000Z, until: 2022-10-10T12:09:01.000Z, type: 'RSA', use: 'sig' }
2022-08-14 11:01:13 INFO:  ssl certificate { check: './cert/fullchain.pem' }
2022-08-14 11:01:13 STATE: ssl certificate validity { days: 57, renew: 'skip' }
2022-08-14 11:01:13 STATE: geoip { city: './geoip/GeoLite2-City.mmdb', asn: './geoip/GeoLite2-ASN.mmdb' }
2022-08-14 11:01:13 INFO:  helmet { frameguard: false, xssFilter: false, dnsPrefetchControl: { allow: true }, noSniff: false, hsts: { maxAge: 15552000, preload: true }, referrerPolicy: { policy: 'no-referrer' }, expectCt: { enforce: true }, contentSecurityPolicy: { directives: { count: [ '10' ] } } }
2022-08-14 11:01:13 INFO:  limiter { interval: 10, tokens: 500 }
2022-08-14 11:01:13 INFO:  compression { brotli: 5 }
2022-08-14 11:01:13 INFO:  proxy { url: 'pimiami.ddns.net', target: 'localhost', port: '10060' }
2022-08-14 11:01:13 INFO:  proxy { url: 'weather.local', target: 'localhost', port: '10060' }
2022-08-14 11:01:13 INFO:  proxy { url: 'stocks.local', target: 'localhost', port: '10040' }
2022-08-14 11:01:13 INFO:  proxy { url: 'pi.hole', target: 'localhost', port: '10050' }
2022-08-14 11:01:13 INFO:  proxy { default: true, target: 'localhost', port: '10020' }
2022-08-14 11:01:13 INFO:  static { paths: [ 'security.txt', 'humans.txt', 'robots.txt', 'git.head', 'sitemap.xml', 'version' ] }
2022-08-14 11:01:13 STATE: server { status: 'listening', address: '::', family: 'IPv6', port: 443 }
2022-08-14 11:01:13 STATE: server { user: { uid: 1000, gid: 1000, username: 'vlado', homedir: '/home/vlado', shell: '/bin/bash' } }
2022-08-14 11:01:13 STATE: noip { hostname: 'pimiami.ddns.net', status: 200, text: 'nochg 159.250.178.205' }
2022-08-14 11:01:15 STATE: monitor { server: { url: 'pimiami.ddns.net', target: 'localhost', port: '10060' }, url: { lookup: false, connect: false, error: '', ready: false, data: false }, target: { lookup: false, connect: false, error: '', ready: false, data: false } }
2022-08-14 11:01:16 STATE: monitor { server: { url: 'weather.local', target: 'localhost', port: '10060' }, url: { lookup: false, connect: false, error: '', ready: false, data: false }, target: { lookup: false, connect: false, error: '', ready: false, data: false } }
2022-08-14 11:01:16 STATE: monitor { server: { url: 'stocks.local', target: 'localhost', port: '10040' }, url: { lookup: false, connect: false, error: '', ready: false, data: false }, target: { lookup: false, connect: false, error: '', ready: false, data: false } }
2022-08-14 11:01:16 STATE: monitor { server: { url: 'pi.hole', target: 'localhost', port: '10050' }, url: { lookup: false, connect: false, error: '', ready: false, data: false }, target: { lookup: false, connect: false, error: '', ready: false, data: false } }
2022-08-14 11:01:17 STATE: monitor { server: { default: true, target: 'localhost', port: '10020' }, url: { lookup: false, connect: false, error: '', ready: false, data: false }, target: { lookup: false, connect: false, error: '', ready: false, data: false } }
2022-08-14 11:01:18 STATE: server { status: 'active', connections: 0, error: null }
```

### Client Requests

```js
2022-08-14 11:05:33 DATA:  { method: 'GET', protocol: 'h2', status: 200, scheme: 'https', host: 'pimiami.ddns.net', url: '/', ip: '::ffff:192.168.0.201', agent: [ 'AppleWebKit/537.36', 'Chrome/104.0.5112.81', 'Safari/537.36', 'Edg/104.0.1293.54' ], client: 'https://pimiami.ddns.net/', device: 'Windows NT 10.0; Win64; x64', mime: 'text/html; charset=utf-8', duration: 550 }
2022-08-14 11:05:33 DATA:  { method: 'GET', protocol: 'h2', status: 200, scheme: 'https', host: 'pimiami.ddns.net', url: '/index.css', ip: '::ffff:192.168.0.201', agent: [ 'AppleWebKit/537.36', 'Chrome/104.0.5112.81', 'Safari/537.36', 'Edg/104.0.1293.54' ], client: 'https://pimiami.ddns.net/index.css', device: 'Windows NT 10.0; Win64; x64', mime: 'text/css; charset=utf-8', duration: 11 }
2022-08-14 11:05:33 DATA:  { method: 'GET', protocol: 'h2', status: 200, scheme: 'https', host: 'pimiami.ddns.net', url: '/assets/leaflet.css', ip: '::ffff:192.168.0.201', agent: [ 'AppleWebKit/537.36', 'Chrome/104.0.5112.81', 'Safari/537.36', 'Edg/104.0.1293.54' ], client: 'https://pimiami.ddns.net/assets/leaflet.css', device: 'Windows NT 10.0; Win64; x64', mime: 'text/css; charset=utf-8', duration: 12 }
2022-08-14 11:05:34 DATA:  { method: 'GET', protocol: 'h2', status: 200, scheme: 'https', host: 'pimiami.ddns.net', url: '/dist/index.js', ip: '::ffff:192.168.0.201', agent: [ 'AppleWebKit/537.36', 'Chrome/104.0.5112.81', 'Safari/537.36', 'Edg/104.0.1293.54' ], client: 'https://pimiami.ddns.net/dist/index.js', device: 'Windows NT 10.0; Win64; x64', mime: 'text/javascript; charset=utf-8', duration: 35 }
2022-08-14 11:05:34 DATA:  { method: 'GET', protocol: 'h2', status: 200, scheme: 'https', host: 'pimiami.ddns.net', url: '/favicon.ico', ip: '::ffff:192.168.0.201', agent: [ 'AppleWebKit/537.36', 'Chrome/104.0.5112.81', 'Safari/537.36', 'Edg/104.0.1293.54' ], client: 'https://pimiami.ddns.net/favicon.ico', device: 'Windows NT 10.0; Win64; x64', mime: 'image/x-icon', duration: 9 }
2022-08-14 11:05:34 DATA:  { method: 'GET', protocol: 'h2', status: 200, scheme: 'https', host: 'pimiami.ddns.net', url: '/dist/pwa-serviceworker.js', ip: '::ffff:192.168.0.201', agent: [ 'AppleWebKit/537.36', 'Chrome/104.0.5112.81', 'Safari/537.36', 'Edg/104.0.1293.54' ], client: 'https://pimiami.ddns.net/dist/pwa-serviceworker.js', device: 'Windows NT 10.0; Win64; x64', mime: 'text/javascript; charset=utf-8', duration: 11 }
2022-08-14 11:05:34 DATA:  { method: 'GET', protocol: 'h2', status: 200, scheme: 'https', host: 'pimiami.ddns.net', url: '/weather.webmanifest', ip: '::ffff:192.168.0.201', agent: [ 'AppleWebKit/537.36', 'Chrome/104.0.5112.81', 'Safari/537.36', 'Edg/104.0.1293.54' ], client: 'https://pimiami.ddns.net/weather.webmanifest', device: 'Windows NT 10.0; Win64; x64', mime: 'application/manifest+json', duration: 5 }
2022-08-14 11:05:34 DATA:  { method: 'GET', protocol: 'h2', status: 200, scheme: 'https', host: 'pimiami.ddns.net', url: '/assets/icons/dash-256.png', ip: '::ffff:192.168.0.201', agent: [ 'AppleWebKit/537.36', 'Chrome/104.0.5112.81', 'Safari/537.36', 'Edg/104.0.1293.54' ], client: 'https://pimiami.ddns.net/assets/icons/dash-256.png', device: 'Windows NT 10.0; Win64; x64', mime: 'image/png', duration: 6 }
```

## Links

- [License](https://github.com/vladmandic/proxy/LICENSE)
- [Contributing Guidelines](https://github.com/vladmandic/proxy/CONTRIBUTING)
- [Security Policy](https://github.com/vladmandic/proxy/SECURITY)
- [Code of Conduct](https://github.com/vladmandic/proxy/CODE_OF_CONDUCT)
- [Change Log](https://github.com/vladmandic/proxy/CHANGELOG.md)




CHANGELOG.md
# @vladmandic/piproxy  

  Version: **2.0.1**  
  Description: **Reverse HTTP/HTTPS/HTTP2 Proxy in NodeJS with SSL Certificate Management, Compression, Security Enforcement and Rate Limiting**  
  
  Author: **Vladimir Mandic <mandic00@live.com>**  
  License: **MIT**  
  Repository: **<https://github.com/vladmandic/piproxy>**  
  
## Changelog
  
### **HEAD -> master** 2022/10/12 mandic00@live.com


### **2.0.1** 2022/08/14 mandic00@live.com

- minor bugfixes
- major rewrite

### **1.1.24** 2022/08/11 mandic00@live.com

- fix loading of service workers

### **1.1.23** 2022/05/25 mandic00@live.com


### **1.1.22** 2022/03/20 mandic00@live.com


### **1.1.21** 2021/12/14 mandic00@live.com


### **1.1.20** 2021/10/02 mandic00@live.com


### **1.1.19** 2021/07/29 mandic00@live.com


### **release: 1.1.8** 2021/04/22 mandic00@live.com


### **1.1.18** 2021/04/21 mandic00@live.com

- fix frame-ancesor for iframes
- reimplement header handling

### **1.1.17** 2021/04/07 mandic00@live.com


### **1.1.16** 2021/03/18 mandic00@live.com


### **1.1.15** 2021/03/13 mandic00@live.com


### **1.1.14** 2021/02/20 mandic00@live.com


### **1.1.13** 2021/02/19 mandic00@live.com


### **1.1.11** 2021/01/29 mandic00@live.com


### **1.1.10** 2021/01/22 mandic00@live.com


### **1.1.9** 2021/01/22 mandic00@live.com

- strict linting

### **1.1.8** 2021/01/03 mandic00@live.com

- merge branch 'master' of https://github.com/vladmandic/piproxy
- merge pull request #3 from vladmandic/dependabot/npm_and_yarn/ini-1.3.8
- bump ini from 1.3.5 to 1.3.8

### **1.1.7** 2020/12/12 mandic00@live.com

- port update
- package update

### **1.1.6** 2020/11/25 mandic00@live.com


### **1.1.5** 2020/11/25 mandic00@live.com


### **2** 2020/10/29 mandic00@live.com


### **1.1.4** 2020/10/28 mandic00@live.com


### **1.1.3** 2020/09/26 mandic00@live.com


### **1.1.2** 2020/09/16 cyan00@gmail.com


### **1.1.1** 2020/09/16 cyan00@gmail.com


### **1.0.24** 2020/09/16 cyan00@gmail.com

- improved error handler

### **1.0.23** 2020/09/11 cyan00@gmail.com


### **1.0.22** 2020/09/11 cyan00@gmail.com

- autorestart on ssl cert update

### **1.0.21** 2020/09/05 cyan00@gmail.com

- fix firefox compression compatibility issue

### **1.0.20** 2020/09/04 cyan00@gmail.com

- add gzip as alternative

### **1.0.19** 2020/09/03 cyan00@gmail.com


### **1.0.18** 2020/08/28 cyan00@gmail.com

- excluded js map files
- monitoring and stats improvements

### **1.0.17** 2020/08/22 cyan00@gmail.com

- reduced dependencies
- tweaks

### **1.0.16** 2020/08/22 cyan00@gmail.com

- econding fix
- compression update

### **1.0.15** 2020/08/19 cyan00@gmail.com

- added geoip db to git lfs
- added

### **1.0.14** 2020/08/19 cyan00@gmail.com

- automatic server monitoring
- added predefined responses

### **1.0.13** 2020/08/19 cyan00@gmail.com

- improved logging

### **1.0.12** 2020/08/16 cyan00@gmail.com


### **1.0.11** 2020/08/14 cyan00@gmail.com

- url query
- enhanced stats
- improved db logging

### **1.0.10** 2020/08/13 cyan00@gmail.com

- initial work on stats page

### **1.0.9** 2020/08/11 cyan00@gmail.com

- fix statuscode
- ip resolving fix

### **1.0.8** 2020/08/10 cyan00@gmail.com

- implemented brotli compression

### **1.0.7** 2020/08/10 cyan00@gmail.com

- priviledge drop

### **1.0.6** 2020/08/08 cyan00@gmail.com

- simplified dependencies

### **1.0.5** 2020/08/06 cyan00@gmail.com

- improved configuration options

### **1.0.4** 2020/08/06 cyan00@gmail.com


### **1.0.3** 2020/08/02 cyan00@gmail.com

- custom error handler

### **1.0.2** 2020/08/02 cyan00@gmail.com

- added helmet protection and rate limiter
- rewrite using native methods

### **1.0.1** 2020/08/01 cyan00@gmail.com

- added secure server redirect

### **0.0.6** 2020/07/29 cyan00@gmail.com


### **release: 1.0.0** 2020/07/19 cyan00@gmail.com


### **0.0.5** 2020/07/19 cyan00@gmail.com

- http2 logging

### **0.0.4** 2020/07/02 cyan00@gmail.com


### **0.0.3** 2020/06/28 cyan00@gmail.com

- noip refresh bugfix
- deleted logs
- added logger with geo location and agent analysis

### **0.0.2** 2020/06/27 cyan00@gmail.com

- readme and systemd file update
- initial commit


src/config.ts
import * as fs from 'fs';
import * as crypto from 'crypto';
import * as log from '@vladmandic/pilogger';
import * as node from '../package.json';

type Redirect = { url?: string, default?: boolean, target: string, port: string };
type Directives = Record<string, string[]>;
type Answers = Record<string, string | object>;

let config = {
  logFile: 'logs/proxy.log' as string, // where to log all actions
  noip: {
    host: [] as string[], // if host array is non-empty piproxy will try to update dynamic ip for all listed hosts
    user: '' as string, // username for noip services
    password: '' as string, // password for noip services
  },
  acme: {
    domains: [] as string[], // if domains array is non-empty piproxy will try to request signed certificate valid for all listed domains from let's encrypt ssl authority
    application: 'piproxy/1.0.0' as string, // signature to use when requesting certificate
    accountFile: './cert/account.json' as string, // path to use
    accountKeyFile: './cert/account.pem' as string, // path to use
    ServerKeyFile: './cert/private.pem' as string, // path to use
    fullChain: './cert/fullchain.pem' as string, // path to use
  },
  ssl: {
    key: '../cert/private.pem' as string, // use this private key when starting https server
    crt: '../cert/fullchain.pem' as string, // use this server certificate when starting https server
  },
  http2: {
    allowHTTP1: true as boolean, // should we allow http1 traffic or force http2 only
    port: 443 as number, // start https server on port
    key: '' as string, // internal
    cert: '' as string, // internal
    secureOptions: crypto.constants.SSL_OP_NO_TLSv1 | crypto.constants.SSL_OP_NO_TLSv1_1, // disallow unsecure protocols
  },
  redirectHTTP: true, // redirect http to https
  redirects: [] as Array<Redirect>, // list of proxy redirects
  limiter: { // if present and vales are non-zero run proxy with per-session rate limiting
    interval: 10 as number, // how long before tokens are reset in seconds
    tokens: 500 as number, // how many requests can a single session request within interval
  },
  compress: 5 as number, // if present and non-zero run brotli compression on outgoing http responses
  monitor: true as boolean, // run monitor function on all source and target hosts/urls listed in redirects
  geoIP: { // use geoip to lookup locations based on ip
    city: './geoip/GeoLite2-City.mmdb' as string, // path to database
    asn: './geoip/GeoLite2-ASN.mmdb' as string, // path to database
  },
  helmet: { // piproxy uses helmet to configure public facing http response options
    frameguard: false as boolean,
    xssFilter: false as boolean,
    dnsPrefetchControl: { allow: true as boolean },
    noSniff: false as boolean,
    hsts: { maxAge: 15552000 as number, preload: true },
    referrerPolicy: { policy: 'no-referrer' },
    expectCt: { enforce: true },
    contentSecurityPolicy: {
      directives: {
        defaultSrc: ["'self'"],
        fontSrc: ["'self'", 'http:', 'https:'],
        imgSrc: ["'self'", 'data:', 'http:', 'https:'],
        mediaSrc: ["'self'", 'data:', 'http:', 'https:'],
        scriptSrc: ["'self'", "'unsafe-inline'", "'unsafe-eval'", 'https:', 'blob:'],
        styleSrc: ["'self'", 'https:', "'unsafe-inline'"],
        connectSrc: ["'self'", 'http:', 'https:', 'data:'],
        workerSrc: ["'self'", 'blob:', 'https:'],
        frameAncestors: ["'self'"],
        'upgrade-insecure-requests': [],
      } as Directives,
    },
  },
  answers: { // predefined answers to common static paths
    'security.txt': 'Contact: mailto:mandic00@live.com\nPreferred-Languages: en\n',
    'humans.txt': '/* TEAM */\nChef: Vladimir Mandic\nContact: mandic00@live.com\nGitHub: https://github.com/vladmandic\n',
    'robots.txt': 'User-agent: *\nDisallow: /private\nCrawl-delay: 10\n',
    'git.head': 'ref: refs/heads/master\n',
    'sitemap.xml': '<?xml version="1.0" encoding="UTF-8"?>\n<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n<url>\n<loc>URL</loc>\n</url>\n</urlset>',
    version: { name: '', version: '' },
  } as Answers,
};

export function merge(...objects) { // helper function: perform deep merge of multiple objects so it allows full inheritance with overrides
  const isObject = (obj) => obj && typeof obj === 'object';
  return objects.reduce((prev, obj) => {
    Object.keys(obj || {}).forEach((key) => {
      const pVal = prev[key];
      const oVal = obj[key];
      if (Array.isArray(pVal) && Array.isArray(oVal)) prev[key] = pVal.concat(...oVal);
      else if (isObject(pVal) && isObject(oVal)) prev[key] = merge(pVal, oVal);
      else prev[key] = oVal;
    });
    return prev;
  }, {});
}

export function init() {
  if (fs.existsSync('config.json')) { // create private secrets file as required
    try {
      const blob = fs.readFileSync('config.json');
      const obj = JSON.parse(blob.toString());
      config = merge(config, obj);
    } catch (err) {
      log.info('Configuration error', { file: 'config.json', err });
    }
    config.answers.version = { name: node.name, version: node.version };
    log.info('Configuration', { file: 'config.json' });
  } else {
    log.info('Configuration missing', { file: 'config.json' });
  }
}

export const get = () => config;

export const set = (obj) => {
  config = merge(config, obj);
};
